pkg/controller/rbac/serviceprincipal/managed.go

```diff
@@ -0,0 +1,158 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package serviceprincipal
+
+import (
+	"context"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/services/graphrbac/1.6/graphrbac"
+	"github.com/Azure/azure-sdk-for-go/services/graphrbac/1.6/graphrbac/graphrbacapi"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/pkg/errors"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/ratelimiter"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplane/provider-azure/apis/rbac/v1alpha1"
+	azure "github.com/crossplane/provider-azure/pkg/clients"
+	azureclients "github.com/crossplane/provider-azure/pkg/clients"
+)
+
+// Error strings.
+const (
+	errNotServicePrincipal    = "managed resource is not an ServicePrincipal"
+	errCreateServicePrincipal = "cannot create ServicePrincipal"
+	errUpdateServicePrincipal = "update ServicePrincipal"
+	errGetServicePrincipal    = "cannot get ServicePrincipal"
+	errDeleteServicePrincipal = "cannot delete ServicePrincipal"
+)
+
+// Setup adds a controller that reconciles ServicePrincipal.
+func Setup(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter, poll time.Duration) error {
+	name := managed.ControllerName(v1alpha1.ServicePrincipalKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+		}).
+		For(&v1alpha1.ServicePrincipal{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.ServicePrincipalGroupVersionKind),
+			// Override default initializers in case to remove NewNameAsExternalName Initializer
+			managed.WithInitializers(),
+			managed.WithConnectionPublishers(),
+			managed.WithExternalConnecter(&connecter{client: mgr.GetClient()}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithPollInterval(poll),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connecter struct {
+	client client.Client
+}
+
+func (c *connecter) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	creds, _, err := azureclients.GetAuthInfo(ctx, c.client, mg)
```

Comment:
Please see the comment on Application connector. If the authorizer returned from `GetAuthInfo` is not suitable, please add a comment explaining the situation.